README.md
# GstCurlHttpSrc

**GstCurlHttpSrc** is a GStreamer URIHandler plugin that supports getting
resources from web servers using HTTP/HTTPS. It uses the curl library to
acheive this, and supports HTTP/1.0, HTTP/1.1 and HTTP/2.0 protocols. (Actual
support is dependent on curl version and configuration)

**GstCurlHttpSrc** implements the URIHandler interface for http:// and https://
URIs, meaning as long as other gstreamer elements or applications use the
URIHandler interface, they need not be changed to support this module as
opposed to using any others.

## Getting Started

**GstCurlHttpSrc** requires [GStreamer](http://gstreamer.freedesktop.org/) 1.x
and any modern version of libcurl. The curl library used may be optionally
compiled against an SSL library such as [OpenSSL](http://www.openssl.org) for
https:// support; and against [nghttp2](http://www.nghttp2.org) for HTTP/2
support.

The software has been developed on Linux Mint 17 (Ubuntu 14.04-alike). This
README has two parts; the **Quick** install and the **Full** install. The Quick
install assumes that you already have a working version of cURL and GStreamer
installed on your machine, in a sensible location. The Full install walks you
through building GStreamer, curl, OpenSSL and nghttp2 from scratch.

In effect, use the Quick guide to just build this module, use the Full guide
if you want everything.

### Quick Guide

#### Download the source

    $ git clone https://github.com/BBC/gst-curlhttpsrc.git

#### Build the source

    $ cd gst-curlhttpsrc
    $ ./autogen.sh
    $ make
    $ make install

Beware that if your GStreamer is installed in `/usr` or somewhere similar, you
will need to be root in order to perform the last command.

#### Run the application

In order to just fetch one file and then dump to the console, you can do the
following:

    $ gst-launch-1.0 curlhttpsrc location=http://example.com/ ! fakesink dump=1

The following assumes that you have a build of GStreamer including all the
necessary plugins in order to play an MPEG-DASH stream. The MPD file in
question has AVC3 video and AAC audio, and comes from
[here](http://rdmedia.bbc.co.uk/dash/ondemand/bbb/). Full details of all
representations and additional MPDs can be found at that link too.

    $ gst-launch-1.0 playbin uri=http://rdmedia.bbc.co.uk/dash/ondemand/bbb/avc3/1/client_manifest-common_init.mpd

#### Default URIHandler

By default, GStreamer uses SoupHTTPSrc as it's URIHandler implementation for
http and https URIs. In order to change back to using SoupHTTPSrc, you can
either remove the CurlHttpSrc modules from `/usr/lib/gstreamer-1.0` or you can
change the element rank so that SoupHTTPSrc takes precidence once again. This
value can be found at the bottom of the gstcurlhttpsrc.c file. Simply change
the line that looks like this:

```
  return gst_element_register (curlhttpsrc, "curlhttpsrc", 500,
      GST_TYPE_CURLHTTPSRC);
```

To this:

```
  return gst_element_register (curlhttpsrc, "curlhttpsrc", 0,
      GST_TYPE_CURLHTTPSRC);
```

### Full Guide

#### Required software

| Package | Project Website | Download |
| ------- | --------------- | -------- |
| GStreamer | http://gstreamer.freedesktop.org/ | [1.4.3 Release](http://gstreamer.freedesktop.org/releases/gstreamer/1.4.3.html) |
| GStreamer Plugins Base | http://gstreamer.freedesktop.org/ | [1.4.3 Release](http://gstreamer.freedesktop.org/releases/gst-plugins-base/1.4.3.html) |
| GStreamer Plugins Good | http://gstreamer.freedesktop.org/ | [1.4.3 Release](http://gstreamer.freedesktop.org/releases/gst-plugins-good/1.4.3.html) |
| Gstreamer Plugins Bad | http://gstreamer.freedesktop.org/ | [1.4.3 Release](http://gstreamer.freedesktop.org/releases/gst-plugins-bad/1.4.3.html) |
| OpenSSL (Optional, for SSL support. Must be 1.1.0 and above for HTTP/2 support) | http://www.openssl.org | git clone https://github.com/openssl/openssl.git |
| nghttp2 (Optional, for HTTP/2 support) | http://nghttp2.org | git clone https://github.com/tatsuhiro-t/nghttp2.git |
| cURL | http://curl.haxx.se | git clone https://github.com/bagder/curl.git |

#### Build preface

It is important to know what you want to do with this software. The intended
usage of this module (at least for the time being) is to show HTTP/2 working.
This requires the usage of some pretty bleeding-edge software that may break,
and as such I wouldn't recommend using it day-to-day or replacing your existing
distribution-supplied versions unless you really want to.

The way I have built this software is to install it in a custom directory in my
home folder, `~/build/target`. The `BUILD_PREFIX` environment variable points
to that directory, with my path being set as `$BUILD_PREFIX/bin:$PATH`, my
`LD_LIBRARY_PATH` variable set as `$BUILD_PREFIX/lib:$PATH` and my
`PKG_CONFIG_PATH` variable set as `$BUILD_PREFIX/lib/pkgconfig`.

If you're fine to replace your existing software (or you don't have any
Distribution supplied ones anyway), then you can ignore any `--prefix` or
`--exec-prefix` options supplied in the build instructions below.

These are all external projects, and therefore the following instructions are
just a guide. The projects are under no control by myself or the BBC and as
such we take no responsibility for the following instructions being correct
as and when you get hold of the software in question.

#### Build GStreamer

GStreamer itself comes in four parts:

* GStreamer Core
    * Contains the core GStreamer framework, as well as top-level applications
such as gst-launch, gst-inspect and playbin.
* GStreamer-Plugins-Base
    * Contains the necessary frameworks to allow loadable plugins
* GStreamer-Plugins-Good
    * Contains a selection of plugins such as SoupHTTPSrc
* GStreamer-Plugins-Bad
    * Contains a selection of plugins with runnable code but lacking something to
stop it from being counted as a "good" plugin, such as:
        * Needs a good code review.
        * Documentation needs improving.
        * No active maintainer.
        * Not being used in a widespread manner.

You'll need to build them in order to minimise any dependency issues.
Thankfully, the build process is the same between each part.

    $ cd gstreamer-1.4.3
    $ ./configure --prefix=$BUILD_PREFIX
    $ make && make install
    $ cd ../gst-plugins-base-1.4.3
    $ ./configure --prefix=$BUILD_PREFIX
    $ make && make install
    $ cd ../gst-plugins-good-1.4.3
    $ # And so on...

#### Build OpenSSL (Optional)

    $ cd openssl
    $ ./config --prefix=$BUILD_PREFIX
    $ make depend
    $ make
    $ make test
    $ make install

#### Build nghttp2 (Optional)

    $ autoreconf -i
    $ ./configure --prefix=$BUILD_PREFIX --exec-prefix=$BUILD_PREFIX
    $ make
    $ make install

#### Build libcurl

    $ ./configure --prefix=$BUILD_PREFIX
    $ make
    $ make install

#### Build CurlHttpSrc

See the Quick Guide section.

## Credits

This plugin contains code derived from the [gst-template](http://cgit.freedesktop.org/gstreamer/gst-template/)
as provided by the GStreamer project.

## License

See the LICENSE file.

## Contributing

If you have a feature request or want to report a bug, we'd be happy to hear
from you. Please either [raise an issue](https://github.com/BBC/gst-curlhttpsrc/issues),
or fork the project and send us a pull request.

## Authors

This software was written by [Sam Hurst](https://github.com/samhurst),
samuelh <at> rd.bbc.co.uk

## Copyright

Copyright 2014 British Broadcasting Corporation



ChangeLog
2014-10-31  Sam Hurst  <samuelh at rd.bbc.co.uk>

    * AUTHORS:
    Added authors file.
    * ChangeLog:
    Added element-specific ChangeLog file, moved gst-template version to
    ChangeLog.template.

    * src/gstcurlhttpsrc.c:
    * src/gstcurlhttpsrc.h:
    * src/curltask.c:
    * src/curltask.h:
    * src/gstcurldefaults.h:
      Initial code checkin.



src/curltask.h
/*
 * curltask.h
 *
 *  Created on: 29 Oct 2014
 *      Author: sam
 */

#ifndef CURLTASK_H_
#define CURLTASK_H_

#include <curl/curl.h>
#include "gstcurldefaults.h"

GST_DEBUG_CATEGORY_STATIC (gst_curl_loop_debug);

#define GSTCURL_ERROR_PRINT(...) gst_debug_log (gst_curl_loop_debug, GST_LEVEL_ERROR, __FILE__, __func__, __LINE__, NULL, __VA_ARGS__)
#define GSTCURL_WARNING_PRINT(...) gst_debug_log (gst_curl_loop_debug, GST_LEVEL_WARNING, __FILE__, __func__, __LINE__, NULL, __VA_ARGS__)
#define GSTCURL_INFO_PRINT(...) gst_debug_log (gst_curl_loop_debug, GST_LEVEL_INFO, __FILE__, __func__, __LINE__, NULL, __VA_ARGS__)
#define GSTCURL_DEBUG_PRINT(...) gst_debug_log (gst_curl_loop_debug, GST_LEVEL_DEBUG, __FILE__, __func__, __LINE__, NULL, __VA_ARGS__)
#define GSTCURL_TRACE_PRINT(...) gst_debug_log (gst_curl_loop_debug, GST_LEVEL_TRACE, __FILE__, __func__, __LINE__, NULL, __VA_ARGS__)

#define gst_curl_setopt_str(handle,type,option) \
	if(option != NULL) { \
		curl_easy_setopt(handle,type,option); \
	} \

#define gst_curl_setopt_int(handle, type, option) \
	if((option > GSTCURL_HANDLE_MIN_##type) && (option < GSTCURL_HANDLE_MAX_##type)) { \
		curl_easy_setopt(handle,type,option); \
	} \

#define gst_curl_setopt_str_default(handle,type,option) \
	if(option == NULL) { \
		curl_easy_setopt(handle,type,GSTCURL_HANDLE_DEFAULT_##type); \
	} \
	else { \
		curl_easy_setopt(handle,type,option); \
	} \

#define gst_curl_setopt_int_default(handle,type,option) \
	if((option < GSTCURL_HANDLE_MIN_##type) || (option > GSTCURL_HANDLE_MAX_##type)) { \
		curl_easy_setopt(handle,type,GSTCURL_HANDLE_DEFAULT_##type); \
	} \
	else { \
		curl_easy_setopt(handle,type,option); \
	} \


struct {

} GstCurlThreadData;

/*
 * Function definitions
 */

#endif /* CURLTASK_H_ */


src/gstcurldefaults.h
/*
 * gstcurldefaults.h
 *
 *  Created on: 31 Oct 2014
 *      Author: sam
 */

#ifndef GSTCURLDEFAULTS_H_
#define GSTCURLDEFAULTS_H_

/*
 * This file contains a list of all the default values used. These are used to
 * initialise an object in its init call.
 *
 * Must all conform to GSTCURL_HANDLE_DEFAULT_##type for macro sillyness in
 * curltask.h, where "type" is the CURLOPT_<something> string.
 */
/* Defaults from http://curl.haxx.se/libcurl/c/curl_easy_setopt.html */
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_URL ""
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_PROXY ""
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_PROXYUSERNAME ""
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_PROXYPASSWORD ""
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_USERAGENT ""
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_FOLLOWLOCATION 1L
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_MAXREDIRS 15L
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_TCP_KEEPALIVE 1L
#define GSTCURL_HANDLE_DEFAULT_CURLOPT_HTTP_VERSION CURL_HTTP_VERSION_1_1

/* Defaults from http://curl.haxx.se/libcurl/c/curl_multi_setopt.html */
#define GSTCURL_HANDLE_DEFAULT_CURLMOPT_PIPELINING 1L
#define GSTCURL_HANDLE_DEFAULT_CURLMOPT_MAXCONNECTS 255L
#define GSTCURL_HANDLE_DEFAULT_CURLMOPT_MAX_HOST_CONNECTIONS 0L
#define GSTCURL_HANDLE_DEFAULT_CURLMOPT_MAX_PIPELINE_LENGTH 5L
#define GSTCURL_HANDLE_DEFAULT_CURLMOPT_MAX_TOTAL_CONNECTIONS 255L

/*
 * Now set acceptable ranges. Defaults can lie outside the range, in which case
 * it is expected that the programmer will use the gst_curl_setopt and not the
 * gst_curl_setopt_default macro, as if the value supplied lies outside of the
 * default range, it won't bother to set it. If the _default macro is used,
 * then the offending value is replaced by the default type above.
 */
#define GSTCURL_HANDLE_MIN_CURLOPT_FOLLOWLOCATION 0L
#define GSTCURL_HANDLE_MAX_CURLOPT_FOLLOWLOCATION 1L
#define GSTCURL_HANDLE_MIN_CURLOPT_MAXREDIRS -1
#define GSTCURL_HANDLE_MAX_CURLOPT_MAXREDIRS 255
#define GSTCURL_HANDLE_MIN_CURLOPT_TCP_KEEPALIVE 0L
#define GSTCURL_HANDLE_MAX_CURLOPT_TCP_KEEPALIVE 1L
#define GSTCURL_HANDLE_MIN_CURLOPT_HTTP_VERSION CURL_HTTP_VERSION_1_0
#define GSTCURL_HANDLE_MAX_CURLOPT_HTTP_VERSION CURL_HTTP_VERSION_2_0

#define GSTCURL_HANDLE_MIN_CURLMOPT_PIPELINING 0L
#define GSTCURL_HANDLE_MAX_CURLMOPT_PIPELINING 1L
#define GSTCURL_HANDLE_MIN_CURLMOPT_MAXCONNECTS 32L
#define GSTCURL_HANDLE_MAX_CURLMOPT_MAXCONNECTS 255L
#define GSTCURL_HANDLE_MIN_CURLMOPT_MAX_HOST_CONNECTIONS 1L
#define GSTCURL_HANDLE_MAX_CURLMOPT_MAX_HOST_CONNECTIONS 127L
#define GSTCURL_HANDLE_MIN_CURLMOPT_MAX_PIPELINE_LENGTH 1L
#define GSTCURL_HANDLE_MAX_CURLMOPT_MAX_PIPELINE_LENGTH 200L
#define GSTCURL_HANDLE_MIN_CURLMOPT_MAX_TOTAL_CONNECTIONS 32L
#define GSTCURL_HANDLE_MAX_CURLMOPT_MAX_TOTAL_CONNECTIONS 255L

#endif /* GSTCURLDEFAULTS_H_ */


src/gstcurlhttpsrc.h
/*
 * GstCurlHttpSrc
 * Copyright 2014 British Broadcasting Corporation - Research and Development
 *
 * Author: Sam Hurst <samuelh@rd.bbc.co.uk>
 *
 * Based on the GstElement template, courtesy of
 * Copyright (C) 2005 Thomas Vander Stichele <thomas@apestaart.org>
 * Copyright (C) 2005 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/*
 * This header file contains definitions only for
 */

#ifndef GSTCURLHTTPSRC_H_
#define GSTCURLHTTPSRC_H_

#include <gst/gst.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <unistd.h>
#include <gst/base/gstpushsrc.h>

#include "curltask.h"

G_BEGIN_DECLS

/* #defines don't like whitespacey bits */
#define GST_TYPE_CURLHTTPSRC \
  (gst_curl_http_src_get_type())
#define GST_CURLHTTPSRC(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_CURLHTTPSRC,GstCurlHttpSrc))
#define GST_CURLHTTPSRC_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_CURLHTTPSRC,GstCurlHttpSrcClass))
#define GST_IS_CURLHTTPSRC(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_CURLHTTPSRC))
#define GST_IS_CURLHTTPSRC_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_CURLHTTPSRC))

/* Because g_param_spec_int requires min/max bounding... */
#define GSTCURL_MIN_REDIRECTIONS -1
#define GSTCURL_MAX_REDIRECTIONS 255
#define GSTCURL_MIN_CONNECTION_TIME 2
#define GSTCURL_MAX_CONNECTION_TIME 60
#define GSTCURL_MIN_CONNECTIONS_SERVER 1
#define GSTCURL_MAX_CONNECTIONS_SERVER 60
#define GSTCURL_MIN_CONNECTIONS_PROXY 1
#define GSTCURL_MAX_CONNECTIONS_PROXY 60
#define GSTCURL_MIN_CONNECTIONS_GLOBAL 1
#define GSTCURL_MAX_CONNECTIONS_GLOBAL 255

#define GSTCURL_DEFAULT_CONNECTION_TIME 30
#define GSTCURL_DEFAULT_CONNECTIONS_SERVER 5
#define GSTCURL_DEFAULT_CONNECTIONS_PROXY 30
#define GSTCURL_DEFAULT_CONNECTIONS_GLOBAL 255

#define GSTCURL_INFO_RESPONSE(x) ((x >= 100) && (x <= 199))
#define GSTCURL_SUCCESS_RESPONSE(x) ((x >= 200) && (x <=299))
#define GSTCURL_REDIRECT_RESPONSE(x) ((x >= 300) && (x <= 399))
#define GSTCURL_CLIENT_ERR_RESPONSE(x) ((x >= 400) && (x <= 499))
#define GSTCURL_SERVER_ERR_RESPONSE(x) ((x >= 500) && (x <= 599))

#define GSTCURL_FUNCTIONTRACE 0

#if GSTCURL_FUNCTIONTRACE
#define GSTCURL_FUNCTION_ENTRY(x) GST_DEBUG_OBJECT(x, "Entering function");
#define GSTCURL_FUNCTION_EXIT(x) GST_DEBUG_OBJECT(x, "Leaving function");
#else
#define GSTCURL_FUNCTION_ENTRY(x)
#define GSTCURL_FUNCTION_EXIT(x)
#endif

typedef struct _GstCurlHttpSrc      		GstCurlHttpSrc;
typedef struct _GstCurlHttpSrcClass 		GstCurlHttpSrcClass;
typedef struct _GstCurlHttpSrcQueueElement	GstCurlHttpSrcQueueElement;

struct _GstCurlHttpSrcClass
{
  GstPushSrcClass parent_class;
};

/*
 * Our instance class.
 */
struct _GstCurlHttpSrc {
	GstPushSrc element;
	/*
	 * Things to tell libcURL about to build up the request message.
	 */
	/* Type		Name					  Curl Option */
	gchar*		uri;					/* CURLOPT_URL */
	gchar*		proxy_uri;				/* CURLOPT_PROXY */
	gchar*		no_proxy_list;			/* CURLOPT_NOPROXY */
	gchar*		proxy_user;				/* CURLOPT_PROXYUSERNAME */
	gchar*		proxy_pass;				/* CURLOPT_PROXYPASSWORD */

	gchar**		cookies;				/* CURLOPT_COOKIELIST */
	gint		number_cookies;
	gchar*		user_agent;				/* CURLOPT_USERAGENT */
	gboolean	allow_3xx_redirect;		/* CURLOPT_FOLLOWLOCATION */
	gint		max_3xx_redirects;		/* CURLOPT_MAXREDIRS */
	gboolean	keep_alive;				/* CURLOPT_TCP_KEEPALIVE */
	/*TODO As the following are all multi options, move these to curl task */
	guint		max_connection_time;	/* */
	guint		max_conns_per_server;	/* CURLMOPT_MAX_HOST_CONNECTIONS */
	guint		max_conns_per_proxy;	/* ?!? */
	guint		max_conns_global;		/* CURLMOPT_MAXCONNECTS */
	/* END multi options */

	/* Some stuff for HTTP/2 */
	enum {
		GSTCURL_HTTP_VERSION_1_0,
		GSTCURL_HTTP_VERSION_1_1,
		GSTCURL_HTTP_VERSION_2_0,
		GSTCURL_HTTP_NOT,	/* For future use, incase not HTTP protocol! */
		GSTCURL_HTTP_VERSION_MAX
	} preferred_http_version;			/* CURLOPT_HTTP_VERSION */

	/*
	 * Mutex for the curl task to hold while it's working.
	 *
	 * It seems we need to add two mutexes, because the calling thread isn't
	 * fast enough to execute before the worker thread attempts to do its thing
	 * and so everything goes pear shaped.
	 */
	GMutex		*mutex;
	GCond		*finished;
	enum {
		GSTCURL_RETURN_NONE,
		GSTCURL_RETURN_DONE,
		GSTCURL_RETURN_BAD_QUEUE_REQUEST,
		GSTCURL_RETURN_TOTAL_ERROR,
		GSTCURL_RETURN_PIPELINE_NULL,
		GSTCURL_RETURN_MAX
	} result;
	CURL*		curl_handle;
	gboolean	end_of_message;

	/*
	 * Response message
	 */
	gchar*		msg;
	guint		len;
	struct {
		gchar*	content_type;
	} headers;

	GstCaps*	caps;
};

struct _GstCurlHttpSrcQueueElement {
	GstCurlHttpSrc				*p;
	GstCurlHttpSrcQueueElement	*next;
	GMutex						*running;
};

static GstCurlHttpSrcQueueElement*	request_queue;
static GMutex*						request_queue_mutex;

static GCond* 	curl_multi_loop_signaller;
static GMutex*	curl_multi_loop_signal_mutex;
static enum {
	GSTCURL_MULTI_LOOP_STATE_WAIT = 0,
	GSTCURL_MULTI_LOOP_STATE_QUEUE_EVENT,
	GSTCURL_MULTI_LOOP_STATE_RUNNING,
	GSTCURL_MULTI_LOOP_STATE_REQUEST_REMOVAL,
	GSTCURL_MULTI_LOOP_STATE_STOP,
	GSTCURL_MULTI_LOOP_STATE_MAX
} curl_multi_loop_signal_state;

static GMutex		GstCurlHttpSrcLoopRefcountMutex;
static guint		GstCurlHttpSrcLoopRefcount;

static GstTask* 	GstCurlHttpSrcLoopTask;
static GRecMutex	GstCurlHttpSrcLoopRecMutex;
static GCond		GstCurlHttpSrcLoopReadyCond;
static GMutex		GstCurlHttpSrcLoopReadyMutex;

static GMutex*			request_removal_mutex;
static GstCurlHttpSrc* 	request_removal_element;

enum
{
  PROP_0,
  PROP_URI,
  PROP_PROXYURI,
  PROP_PROXYUSERNAME,
  PROP_PROXYPASSWORD,
  PROP_COOKIES,
  PROP_USERAGENT,
  PROP_REDIRECT,
  PROP_MAXREDIRECT,
  PROP_KEEPALIVE,
  PROP_CONNECTIONMAXTIME,
  PROP_MAXCONCURRENT_SERVER,
  PROP_MAXCONCURRENT_PROXY,
  PROP_MAXCONCURRENT_GLOBAL,
  PROP_HTTPVERSION,
  PROP_MAX
};

static curl_version_info_data*	gst_curl_http_src_curl_capabilities;
static gfloat					pref_http_ver;
static gchar*					gst_curl_http_src_default_useragent;

/*
 * Function Definitions
 */
/* Gstreamer generic element functions */
static void gst_curl_http_src_class_init (GstCurlHttpSrcClass * klass);
static void gst_curl_http_src_set_property (GObject * object, guint prop_id,
		const GValue * value, GParamSpec * pspec);
static void gst_curl_http_src_get_property (GObject * object, guint prop_id,
		GValue * value, GParamSpec * pspec);
static void gst_curl_http_src_init (GstCurlHttpSrc * source);
static GstFlowReturn gst_curl_http_src_create (GstPushSrc * psrc,
		GstBuffer ** outbuf);
static GstFlowReturn
gst_curl_http_src_handle_response(GstCurlHttpSrc *src, GstBuffer **buf);
static gboolean gst_curl_http_src_negotiate_caps (GstCurlHttpSrc *src);
static GstStateChangeReturn gst_curl_http_src_change_state (
		GstElement * element, GstStateChange transition);

/* URI Handler functions */
static void gst_curl_http_src_uri_handler_init (gpointer g_iface,
		gpointer iface_data);
static guint gst_curl_http_src_urihandler_get_type (GType type);
static const gchar *const * gst_curl_http_src_urihandler_get_protocols (GType type);
static gchar * gst_curl_http_src_urihandler_get_uri (GstURIHandler * handler);
static gboolean gst_curl_http_src_urihandler_set_uri (GstURIHandler * handler,
		const gchar * uri, GError ** error);

/* GstTask functions */
static void gst_curl_http_src_curl_multi_loop(gpointer thread_data);
static CURL* gst_curl_http_src_create_easy_handle(GstCurlHttpSrc *s);
static gboolean gst_curl_http_src_make_request(GstCurlHttpSrc *s);
static inline void gst_curl_http_src_destroy_easy_handle(CURL* handle);
static size_t gst_curl_http_src_get_header(void *header, size_t size,
		size_t nmemb, GstCurlHttpSrc *s);
static size_t gst_curl_http_src_get_chunks(void *chunk, size_t size,
		size_t nmemb, GstCurlHttpSrc *s);
static gboolean gst_curl_http_src_signal_finished(CURL* handle,	gint reason);
static void inline gst_curl_http_src_recurse_queue_cleanup(
		GstCurlHttpSrcQueueElement* element, gint reason);
static void gst_curl_http_src_request_remove(GstCurlHttpSrc* src);

static gboolean gst_curl_try_mutex(GMutex* gmutex);
static char* gst_curl_http_src_strcasestr(const char* haystack, const char* needle);

G_END_DECLS

#endif /* GSTCURLHTTPSRC_H_ */
